nmigen_boards/quickfeather.py

```diff
@@ -0,0 +1,91 @@
+
+
+class QuickfeatherPlatform(QuicklogicPlatform):
+    device      = "ql-eos-s3_wlcsp"
+    part        = "PU64"
+    default_clk = "sys_clk0"
+    connectors = [
+        Connector("J", 2, "- 28 22 21 37 36 42 40 7 2 4 5"),
+        Connector("J", 3, "- 8 9 17 16 20 6 55 31 25 47 - - - - 41"),
+        Connector("J", 8, "27 26 33 32 23 57 56 3 64 62 63 61 59 - - -"),
+    ]
+    resources   = [
+        # This is a placeholder resource since the board utilizes
+        # default internal SoC clock.
+        Resource("sys_clk0", 0, Pins("63", dir="i"), Clock(10e6)),
+
+        *ButtonResources(pins="62"),
+
+        RGBLEDResource(0, r="34", g="39", b="38"),
+
+        UARTResource(0,
+            rx="9", tx="8",
+        ),
+
+        Resource("spi_master", 0,
+            Subsignal("clk", Pins("20", dir="i")),
+            Subsignal("miso", Pins("17", dir="o")),
+            Subsignal("mosi", Pins("16", dir="i")),
+            Subsignal("ss", Pins("11", dir="i")),
+            Subsignal("cs2", Pins("28", dir="i")),
+            Subsignal("cs3", Pins("18", dir="i")),
+        ),
+
+        Resource("spi_slave", 0,
+            Subsignal("clk", Pins("40", dir="i")),
+            Subsignal("miso", Pins("42", dir="o")),
+            Subsignal("mosi", Pins("36", dir="i")),
+            Subsignal("cs_n", Pins("37", dir="i")),
+        ),
+
+        Resource("i2c", 0,
+            Subsignal("scl", Pins("4", dir="io")),
+            Subsignal("sda", Pins("5", dir="io")),
+        ),
+        Resource("i2c", 1,
+            Subsignal("scl", Pins("22", dir="io")),
+            Subsignal("sda", Pins("21", dir="io")),
+        ),
+        DirectUSBResource(0, d_p="10", d_n="14"),
+        Resource("swd", 0,
+            Subsignal("clk", Pins("54", dir="io")),
+            Subsignal("io",  Pins("53", dir="io")),
+        ),
+    ]
+
+    # This programmer requires OpenOCD with support for eos-s3:
+    # https://github.com/antmicro/openocd/tree/eos-s3-support
+    def toolchain_program(self, products, name):
+        openocd = os.environ.get("OPENOCD", "openocd")
+        gdb = os.environ.get("GDB", "gdb")
+        with products.extract("{}.bit".format(name)) as bitstream_filename:
+            bitstream_folder = os.path.dirname(bitstream_filename)
+            top_path = bitstream_folder + "/top.cfg"
+            subprocess.call(["python", "-m", "quicklogic_fasm.bitstream_to_openocd", bitstream_filename, top_path])
+            try:
+                openocd_proc = subprocess.Popen(["openocd", "-s", "tcl",
+                                                 "-f", "interface/ftdi/antmicro-ftdi-adapter.cfg",
+                                                 "-f", "interface/ftdi/swd-resistor-hack.cfg",
+                                                 "-f", "board/quicklogic_quickfeather.cfg",
+                                                 "-f", top_path])
+                gdb_commands = ["tar rem :3333", "monitor reset halt", "monitor load_bitstream"]
```

Comment:
Could you explain how the flashing process works (or link some documentation)? This approach seems mostly fine but I'd like to understand what `load_bitstream` does.

Reply:
Sure, `load_bitstream` is a generated openocd function which consists of a series of writes to registers - there are configuration writes and a bitstream in form of `mww` writes. Here is the [script](https://github.com/QuickLogic-Corp/quicklogic-fasm/blob/master/quicklogic_fasm/bitstream_to_openocd.py) which generates this function. The script produces `top.cfg` as stated [here](https://github.com/nmigen/nmigen-boards/pull/117/files/83fe77ef814e12dd7d17fbf8b6ebb5c1aa707462#diff-d5dc0d5c389715684c33091ab150cd123e57b5c09e5b9b851ee4ea2782c7fc29R72-R73) and as it is passed to openocd, it can be used with the `monitor` command. Additionally to make pins work with the design, the iomux must be configured also. I can see that the toolchain produces a file `top_iomux.openocd` which is again a series of `mww` writes and to make the Blinky example actually work, these commands should be executed in GDB also. The question is whether we should assume that the user is going to do that manually or we would need some additional code to merge `top_iomux.openocd` file into `top.cfg` file?

Reply:
Do I understand it correctly that to program the bitstream, this code currently runs openocd (in the background), then runs gdb that connects to this openocd instance, and then issues two monitor commands? Is there a reason you need gdb at all in this combo?

As far as I can see so far, the gdb invocation can be safely replaced with e.g. command line options `-c 'reset halt' -c 'load_bitstream'` in the openocd invocation. This would greatly simplify the script and avoid issues with some other process on the developer's machine, maybe another openocd instance, using the port 3333.

> I can see that the toolchain produces a file `top_iomux.openocd` which is again a series of `mww` writes and to make the Blinky example actually work, these commands should be executed in GDB also. The question is whether we should assume that the user is going to do that manually or we would need some additional code to merge top_iomux.openocd file into top.cfg file?

We should definitely do this automatically so that the `toolchain_program` function is useful standalone. The same considerations apply here as above.

Reply:
Yea, you're right. I removed GDB from the flow and replaced it with direct commands to OpenOCD. I also added some code to merge openocd scripts (bitstream and iomux) into one.